amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpSinkStage.scala

```diff
@@ -48,32 +49,136 @@ final class AmqpSinkStage(settings: AmqpSinkSettings)
 
       override def whenConnected(): Unit = {
         val shutdownCallback = getAsyncCallback[ShutdownSignalException] { ex =>
+          promise.failure(ex)
           failStage(ex)
         }
-        channel.addShutdownListener(new ShutdownListener {
-          override def shutdownCompleted(cause: ShutdownSignalException): Unit =
-            shutdownCallback.invoke(cause)
-        })
+        channel.addShutdownListener(
+          new ShutdownListener {
+            override def shutdownCompleted(cause: ShutdownSignalException): Unit =
+              shutdownCallback.invoke(cause)
+          }
+        )
         pull(in)
       }
 
-      setHandler(in,
+      setHandler(
+        in,
         new InHandler {
-        override def onPush(): Unit = {
-          val elem = grab(in)
-          channel.basicPublish(
-            exchange,
-            routingKey,
-            elem.mandatory,
-            elem.immediate,
-            elem.props.orNull,
-            elem.bytes.toArray
-          )
-          pull(in)
-        }
-      })
 
-    }
+          override def onUpstreamFailure(ex: Throwable): Unit = {
+            promise.failure(ex)
+            super.onUpstreamFailure(ex)
+          }
+
+          override def onUpstreamFinish(): Unit = {
+            promise.success(Done)
+            super.onUpstreamFinish()
+          }
+
+          override def onPush(): Unit = {
+            val elem = grab(in)
+            channel.basicPublish(
+              exchange,
+              routingKey,
+              elem.mandatory,
+              elem.immediate,
+              elem.props.orNull,
+              elem.bytes.toArray
+            )
+            pull(in)
+          }
+        }
+      )
+    }, promise.future)
```

Comment:
Could be a good idea to put a guard in there (`postStop`) try-failing the promise with an exception and "stage stopped unexpectedly" so that it does not happen silently.